### ISSUE TITLE
Edges do not render initially, 'Couldn't create edge for source handle id: "null"' when using ELK

### ISSUE THREAD

**Issue by lucas-amberg** [xyflow/xyflow#4537]
### What platform were you using when you found the bug?

- React Flow / Svelte Flow version: 12.0.4
- Browser and version: Chrome 127.0.6533.100
- OS and version: Windows 11 23H2

### Describe the Bug

When using ELK js, all of my nodes render in a tree format as expected, and the edges update properly using `setEdges`, but they do not render and display the following error:

![image_2024-08-12_100021532](https://github.com/user-attachments/assets/5b594b55-30be-4657-992e-9395b34393cf)

_(4 edges logged, expected, but not rendering)_,

However when I make a change to my code and save, the edges render propertly 🤔 

![image_2024-08-12_100149815](https://github.com/user-attachments/assets/4ed03b4d-bbb9-4a4e-baa8-ac3dfc3f50cc)

And this only happens when using the formattedNodes from ELK, if I use my nodes from before formatting, it does work as expected, just without the tree formatting I look to achieve from ELK. Just to note, neither the ELK formatted edges or the initial edges work in this scenario, both throw the same error, so it is likely related to ELK and the Nodes rather than the edges.

![image_2024-08-12_100651656](https://github.com/user-attachments/assets/e15aa412-f1ef-49ef-ae27-80994e9c24c2)

## Code:

Here is some of the code, if anyone needs any more let me know.

Here is my useEffect after fetching the data
```typescript title="useEffect"
        let orgMemberRelationships: Edge[] = [];

        orgMemberSupervisors.forEach((supervisor) => {
            Object.values(supervisor["-SUPERVISOR_TO->User"]!).forEach(
                (supervised) => {
                    orgMemberRelationships.push({
                        id: supervisor.nodeId + "->" + supervised.nodeId,
                        source: supervisor.nodeId,
                        target: supervised.nodeId,
                        sourceHandle: 'a',
                        ...edgeStyles,
                    });
                },
            );
        });

        const orgMemberNodes = orgDataArray.map((orgMember, idx) => {
            const depth = Math.floor(Math.log2(idx + 1.5));
            const positionInLevel = idx - (Math.pow(2, depth) - 1);

            return {
                id: orgMember.nodeId,
                type: "orgMember",
                data: orgMember,
                position: {
                    x: 100 + positionInLevel * 400 * Math.pow(-1, idx),
                    y: 400 + depth * 400,
                },
            };
        });

        // Use an async function inside useEffect to handle the Promise
        const applyLayout = async () => {
            try {
                const { nodes: layoutedNodes, edges: layoutedEdges } =
                    await getLayoutedGraph(
                        orgMemberNodes,
                        orgMemberRelationships,
                    );
                    setNodes(layoutedNodes as Node[])
                    console.log(layoutedNodes)
            } catch (error) {
                console.error("Error applying layout:", error);
            }
        };

        

        applyLayout().then(() => {
            console.log('applying');
            setTimeout(() => {
                applyEdges(orgMemberRelationships);
            }, 3000);
        });
        
```

My ELK formatting function

``` typescript

const elk = new ELK();

const elkOptions = {
    "elk.algorithm": "layered",
    "elk.layered.spacing.nodeNodeBetweenLayers": "100",
    "elk.spacing.nodeNode": "80",
    "elk.direction": "DOWN",
};


const getLayoutedGraph = async (nodes: Node[], edges: Edge[]) => {
    const graph = {
        id: "root",
        layoutOptions: elkOptions,
        children: nodes.map((node) => ({
            ...node,
            // Adjust the target and source handle positions based on the layout
            // direction.

            // Hardcode a width and height for elk to use when layouting.
            width: 350,
            height: 250,
            sourcePosition: "bottom",
            targetPosition: "top",
        })),
        edges: edges.map((edge) => {
            return {
                ...edge,
                sourceHandle: 'a',
                sources: [edge.source],
                targets: [edge.target],
            };
        }),
        
    };

    const layoutedGraph = await elk.layout(graph);

    if (layoutedGraph) {
        return {
            nodes: layoutedGraph.children!.map((node) => ({
                ...node,
                // React Flow expects a position property on the node instead of `x`
                // and `y` fields.
                position: { x: node.x, y: node.y },
            })),

            edges: layoutedGraph.edges,
        };
    } else {
        return { nodes, edges };
    }
};
```

Additionally, I've tried manually setting the node handle Ids and that did not make any difference.

### Steps to reproduce the bug or issue

1. Create a set of nodes and edges
2. Format using ELK js
3. Call setNodes and setEdges

### Expected behavior

I expected all of the edges to render properly when setEdges is called, not with a change to the code.

**Comment by peterkogo** [comment]
Can show us your edges that you are ultimately passing to React Flow, that produces this error? Also, is it possible you are only rendering your <Handle /> components in a delayed fashion?

**Comment by bcakmakoglu** [comment]
Your code suggests that you assign `a` as the source handle id to edges:
```ts
orgMemberRelationships.push({
    id: supervisor.nodeId + "->" + supervised.nodeId,
    source: supervisor.nodeId,
    target: supervised.nodeId,
    sourceHandle: 'a', // <---- source handle id `a` is assigned here
    ...edgeStyles,
});
```

Any chance your `<Handle>` component uses a different id than `a` or possibly has *no id* assigned at all?

**Comment by lucas-amberg** [comment]
Yes @bcakmakoglu , I assigned the source <Handle? the id of "a" and the edges that id because I was assuming leaving it unassigned was what was throwing that "null" error

**Comment by lucas-amberg** [comment]
> Can show us your edges that you are ultimately passing to React Flow, that produces this error? Also, is it possible you are only rendering your components in a delayed fashion?

Not sure if this is what you are looking for, but here are the edges after they are set using setEdges and logged on the re render

![image_2024-08-12_103213423](https://github.com/user-attachments/assets/3c0e2a6b-7d2f-4072-9392-9c93d0369b69)

**Comment by bcakmakoglu** [comment]
>  I assigned the source <Handle? the id of "a" and the edges that id because I was assuming leaving it unassigned was what was throwing that "null" error

I didn't quite understand that. Can you show us an excerpt of your `<Handles>`? Both source and target ones?

Leaving it unassigned is perfectly fine, given that you only use *one* `<Handle>` of each type (`source` or `target`).
You only ever need an id if there are multiple handles of the same type.

**Comment by lucas-amberg** [comment]
> I didn't quite understand that. Can you show us an excerpt of your `<Handles>`? Both source and target ones?
> 
> Leaving it unassigned is perfectly fine, given that you only use _one_ `<Handle>` of each type (`source` or `target`). You only ever need an id if there are multiple handles of the same type.

Source Handle:
```typescript
<Handle
      style={{ background: "transparent" }}
      type="source"
      id="a"
      position={Position.Bottom}
  />
```

Target handle:

```typescript
<Handle
      type="target"
      style={{ background: "transparent" }}
      id="b"
      position={Position.Top}
  />
```

**Comment by bcakmakoglu** [comment]
Mh... Any conditions on rendering these Handles?
Also any luck if you just remove these ids since they're unnecessary anyway? ^^

**Comment by lucas-amberg** [comment]
There's no conditions, they always render no matter what. No luck with removing the Ids unfortunately, I added those in there hoping to fix this issue so they aren't the cause. I managed to swap out ELK for dagre and i'm getting better results so unless you guys want to keep iterating on this we can move on from it, thank you for the help 🫡

**Comment by moklick** [comment]
I would really like to understand why this is not working for you. Any chance we could check out the repo or could you create a codesandbox?

**Comment by lucas-amberg** [comment]
@moklick I'll try to make a codesandbox if I get the chance

**Comment by moklick** [comment]
that sounds great @lucas-amberg

**Comment by roppa** [comment]
Not sure if it is the same thing, but for me I used "input" instead of "default" on my nodes. The error (same as what you got) and DX really didn't help. Had to pull everything apart, sore eyes. I'm getting too old for this sh*t. And yes, I am a noob with ReactFlow

**Comment by moklick** [comment]
Please re-open with a sandbox or repo if the issue still occurs.

**Comment by gstark** [comment]
@lucas-amberg  @moklick I seem to be having a similar issue. I'm loading node and edge data and then setting those in react state at the same time. If I do that I do not see edges rendered. However, if I do something like this:

```
      setNodes(parsedState.nodes);
      setTimeout(() => {
        setEdges(parsedState.edges);
      }, 1000);
```

I *do* see the edges rendered. It is almost like the nodes cannot be "seen" by the edges, don't work unless the nodes have been processed/rendered. I will continue investigating this and see if I can make a reproducible example.

**Comment by gstark** [comment]
@lucas-amberg @moklick After researching more, I discovered that my `Handle` components were conditionally rendered in the various nodes. This explains why my setTimeout hack above worked, albeit only occasionally.

I solved my issue by either:
- Ensuring that at least the `<Handle ` is rendered unconditionally in my component
or
- Using `useUpdateNodeInternals` when the conditions changed and the handles were rendered. This is documented in the online help for `008`: https://reactflow.dev/learn/troubleshooting#008. I did not consider that I was "adding or removing handles," but it is clear now that conditionally rendering the handles is the same situation.
